system/lib/wasmfs/wasmfs.cpp

```diff
@@ -76,16 +78,78 @@ __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
 
   auto lockedFile = file->locked();
 
-  __wasi_size_t num = 0;
+  off_t offset = lockedOpenFile.position();
   for (size_t i = 0; i < iovs_len; i++) {
     const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+    size_t len = iovs[i].buf_len;
 
-    lockedFile.write(buf, len);
-    num += len;
+    // Check if the sum of the buf_len values overflows a uint64_t (file size).
+    if (__wasmfs_offset_t(offset) + __wasmfs_offset_t(len) <
+        __wasmfs_offset_t(offset)) {
+      return __WASI_ERRNO_INVAL;
+    }
+
+    // Check if buf_len specifies a positive length buffer but buf is a
+    // null pointer
+    if (!buf && len > 0) {
+      return __WASI_ERRNO_INVAL;
+    }
+
+    lockedFile.write(buf, len, offset);
+    offset += len;
+  }
+
+  *nwritten = offset - lockedOpenFile.position();
+  lockedOpenFile.position() = offset;
+
+  return __WASI_ERRNO_SUCCESS;
+}
+
+__wasi_errno_t __wasi_fd_read(__wasi_fd_t fd,
+                              const __wasi_iovec_t* iovs,
+                              size_t iovs_len,
+                              __wasi_size_t* nread) {
+  auto openFile = FileTable::get()[fd];
+
+  if (!openFile) {
+    return __WASI_ERRNO_BADF;
   }
-  *nwritten = num;
 
+  auto lockedOpenFile = openFile.locked();
+  auto* file = lockedOpenFile.getFile()->dynCast<DataFile>();
+
+  // If file is nullptr, then the file was not a DataFile.
+  if (!file) {
+    return __WASI_ERRNO_ISDIR;
```

Comment:
is ISDIR the right result if it is a symlink, which is the other possibility? (if not, a TODO here would be good)

Reply:
Yep, will add a TODO here when symlinks are added.